File: 2830-maximize-the-profit-as-the-salesman/2830-maximize-the-profit-as-the-salesman.cpp
class Solution {
private:
    int findMaxProfit(int index, vector<vector<int>>& offers, vector<int>& dp) {
        if(index == offers.size()) return 0;
        if(dp[index] != -1) return dp[index];
        
        int not_sell = findMaxProfit(index + 1, offers, dp);
        int next_index = offers.size();
        
        int low = index + 1;
        int high = offers.size() - 1;
        
        while(low <= high) {
            int mid = (low + high) / 2;
            if(offers[mid][0] > offers[index][1]) {
                high = mid - 1;
                next_index = mid;
            } 
            
            else low = mid + 1;
        }

        int sell = offers[index][2] + findMaxProfit(next_index, offers, dp);
        
        return dp[index] = max(sell, not_sell);
    }
    
public:
    int maximizeTheProfit(int n, vector<vector<int>>& offers) {
        int m = offers.size();
        sort(begin(offers), end(offers));
        
        vector<int> dp (m, -1);
        return findMaxProfit(0, offers, dp);
    }
};

File: 2830-maximize-the-profit-as-the-salesman/2830-maximize-the-profit-as-the-salesman.java
class Solution {
    private int findMaxProfit(int index, List<List<Integer>> offers, int[] dp) {
        if (index == offers.size()) return 0;
        if (dp[index] != -1) return dp[index];

        int notSell = findMaxProfit(index + 1, offers, dp);
        int nextIndex = offers.size();

        int low = index + 1;
        int high = offers.size() - 1;

        while (low <= high) {
            int mid = (low + high) / 2;
            if (offers.get(mid).get(0) > offers.get(index).get(1)) {
                high = mid - 1;
                nextIndex = mid;
            } 
            
            else low = mid + 1;
        }

        int sell = offers.get(index).get(2) + findMaxProfit(nextIndex, offers, dp);

        return dp[index] = Math.max(sell, notSell);
    }

    public int maximizeTheProfit(int n, List<List<Integer>> offers) {
        int m = offers.size();
        Collections.sort(offers, Comparator.comparingInt(o -> o.get(0)));

        int[] dp = new int[m];
        Arrays.fill(dp, -1);

        return findMaxProfit(0, offers, dp);
    }
}
